src/db.c

```diff
@@ -1001,13 +1021,6 @@ void scanGenericCommand(client *c, robj *o, unsigned long cursor) {
             }
         }
 
-        /* Filter an element if it isn't the type we want. */
-        if (!filter && o == NULL && typename){
-            robj* typecheck = lookupKeyReadWithFlags(c->db, kobj, LOOKUP_NOTOUCH);
```

Comment:
note that another side effect of this change is that we don't attempt to expire keys that didn't match the type (or pattern).
i.e. both lookupKeyReadWithFlags and expireIfNeeded below will do that.
but we don't want to do that from within the scan callback.
some people use SCAN in order to force expiration on keys, but i suppose they won't use any filters, so that's probably ok.

Reply:
ohh, i see you added a test (to validate that the matched ones do get expired).
we'll, we'll also need to document the behaviour change about the non-matched ones, in the PR top comment, so that we mention that in the release notes.

Reply:
yes, it will have the side effect.  I think it would be better to expire the keys only when matched, whether it is match pettern or match type. And the current strategy is very confusing: if pattern not mached, the key will not attempt to expire, while when type not mached, the key still attempt to expire.

Reply:
i agree. just saying we must clearly document the behavior change in the top comment and release notes.